ctapipe_io_lst/__init__.py

```diff
@@ -256,44 +278,54 @@ def create_subarray(self, tel_id=1):
         # LSTs telescope position taken from MC from the moment
         tel_positions = {tel_id: [50., 50., 16] * u.m}
 
-        subarray = SubarrayDescription("LST1 subarray")
-        subarray.tel_descriptions = tel_descriptions
-        subarray.positions = tel_positions
-        subarray.tel[tel_id] = lst_tel_descr
+        subarray = SubarrayDescription(
+            name=f"LST-{tel_id} subarray",
+            tel_descriptions=tel_descriptions,
+            tel_positions=tel_positions,
+        )
 
         return subarray
 
     def _generator(self):
 
         # container for LST data
-        self.data = LSTDataContainer()
-        self.data.meta['input_url'] = self.input_url
-        self.data.meta['max_events'] = self.max_events
-        self.data.meta['origin'] = 'LSTCAM'
+        array_event = LSTArrayEventContainer()
+        array_event.meta['input_url'] = self.input_url
+        array_event.meta['max_events'] = self.max_events
+        array_event.meta['origin'] = 'LSTCAM'
 
         # fill LST data from the CameraConfig table
-        self.fill_lst_service_container_from_zfile()
+        array_event.lst.tel[self.tel_id].svc = self.fill_lst_service_container()
 
         # initialize general monitoring container
-        self.initialize_mon_container()
+        self.initialize_mon_container(array_event)
 
         # loop on events
-        for count, event in enumerate(self.multi_file):
+        for count, zfits_event in enumerate(self.multi_file):
+            array_event.count = count
+            array_event.index.event_id = zfits_event.event_id
+            array_event.index.obs_id = self.obs_ids[0]
 
-            self.data.count = count
-            self.data.index.event_id = event.event_id
-            self.data.index.obs_id = self.obs_id
+            # fill general R0 data
+            self.fill_r0_container(array_event, zfits_event)
 
             # fill specific LST event data
-            self.fill_lst_event_container_from_zfile(event)
+            self.fill_lst_event_container(array_event, zfits_event)
+
+            # fill trigger info (needs r0 and lst specifics)
+            self.fill_trigger_info(array_event, zfits_event)
 
             # fill general monitoring data
-            self.fill_mon_container_from_zfile(event)
+            self.fill_mon_container(array_event, zfits_event)
+            self.fill_pointing_info(array_event)
 
-            # fill general R0 data
-            self.fill_r0_container_from_zfile(event)
+            if self.r0_r1_calibrator.drs4_pedestal_path.tel[self.tel_id] is not None:
+                self.r0_r1_calibrator.calibrate(array_event)
 
-            yield self.data
+                # tagging flat field events needs r1
+                self.tag_flatfield_events(array_event)
```

Comment:
Hi,
for the calibration calculator code we need the correct event tagging (pedestal and flat field tagging) before the dc_to_pe conversion More precisely, we need a drs4 corrected waveform, with corrected event type, that is still not calibrated (as the lstchain R1 level). Personally I would split the (now a bit too long "calibrate" function) in two functions: a drs4 correction  function (including the baseline, time lapse, interpolation, time correction and  event tagging) and a calibrate (including gain selection, dc_to pe, ff time correction). The calibration code will use only the first function.

Reply:
Ok, will split it into DRS4 corrections and calibration code and do the tagging after applying DRS4. Don't know when I will be able to do that but should be possible this week.